go/pkg/tool/tool.go

```diff
@@ -329,39 +329,41 @@ func (c *Client) DownloadDirectory(ctx context.Context, rootDigest, path string)
 // UploadStats contains various metadata of a directory upload.
 type UploadStats struct {
 	rc.TreeStats
-	RootDigest              digest.Digest
-	TotalNumBlobs           int64
-	NumCacheMisses          int64
-	CompressedBytesUploaded int64
-	CacheMissesBytes        int64
+	RootDigest       digest.Digest
+	TotalNumBlobs    int64
+	NumCacheMisses   int64
+	BytesTransferred int64
+	CacheMissesBytes int64
+	Error            string
 }
 
 // UploadDirectory uploads a directory from the specified path as a Merkle-tree to the remote cache.
 func (c *Client) UploadDirectory(ctx context.Context, path string) (*UploadStats, error) {
 	log.Infof("Computing Merkle tree rooted at %s", path)
 	root, blobs, stats, err := c.GrpcClient.ComputeMerkleTree(ctx, path, "", "", &command.InputSpec{Inputs: []string{"."}}, filemetadata.NewNoopCache())
 	if err != nil {
-		return nil, err
+		return &UploadStats{Error: err.Error()}, err
+	}
+	us := &UploadStats{
```

Comment:
I'd say to move this before the condition at line 344 so the progress can be registered. It's useful to know if any progress was made or not.

Reply:
That's not how ComputeMerkeTree [works](https://github.com/bazelbuild/remote-apis-sdks/blob/46b7334fa1151e051515d606ad849e2856ea7210/go/pkg/client/tree.go#L412): it follows the pattern that we're breaking here, which is the return value of everything else is empty when a non-nil error is returned. So we don't have any partial progress. Ditto below.
We can change that contract, but I think it's better done separately.

Reply:
Sounds good for separate PR. Thanks!